### ISSUE TITLE
Insert mode with obscured input fields

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#253]
On some pages, input fields are obscured by div's or similiar things.

Some examples:
- Input fields on http://www.sbb.ch/
- Codemirror (e.g. http://p.cmpl.cc/)
- Google Docs editor
- Ace (e.g. http://ace.c9.io/)
- [Yahoo JP login](https://login.yahoo.co.jp/config/login)
- Jupyter notebook (selects `<span class="cm-string">`)
- [CKeditor](http://ckeditor.com/) (separate HTML `body` in an iframe)
- Slack (also see https://github.com/qutebrowser/qutebrowser/issues/2215#issuecomment-459623184)

We currently implement workarounds for this cases by matching on div-classes, but we have to investigate if this is a good algorithm:
- Find all children elements of the clicked elements which could be input elements
- Check the geometry for each one, and if it's completely inside (i.e. hidden by) the clicked element, treat it as insert mode.

**Comment by The-Compiler** [comment]
Currently the TinyMCE editor (used e.g. in the Orange Community) also doesn't trigger insert mode.

There, `webelem.WebElementWrapper(elem.lastChild()).is_editable()` is `True` - I wonder if that'd also work for other similiar cases, and if it'd introduce any false-positives.

The raw element in this case seems to be a complete HTML document (?!?):

```html
<html><head><style id="mceDefaultStyles" type="text/css">.mce-content-body div.mce-resizehandle {position: absolute;border: 1px solid black;background: #FFF;width: 5px;height: 5px;z-index: 10000}.mce-content-body .mce-resizehandle:hover {background: #000}.mce-content-body img[data-mce-selected], hr[data-mce-selected] {outline: 1px solid black;resize: none}.mce-content-body .mce-clonedresizable {position: absolute;outline: 1px dashed black;opacity: .5;filter: alpha(opacity=50);z-index: 10000}.mce-content-body .mce-resize-helper {background: #555;background: rgba(0,0,0,0.75);border-radius: 3px;border: 1px;color: white;display: none;font-family: sans-serif;font-size: 12px;white-space: nowrap;line-height: 14px;margin: 5px 10px;padding: 5px;position: absolute;z-index: 10001}\r\n</style><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><link type="text/css" rel="stylesheet" href="http://community.orange.ch/html/js/lib/tinymce/4.1.6/skins/lightgray/content.min.css"><link type="text/css" rel="stylesheet" href="http://orangech.i.lithium.com/skins/383656/B702878441EFE1264294752444A24929/orange_community_skin-1156689385.css"></head><body id="tinymce" class="mce-content-body lia-message-editor" data-id="tinyMceEditor" contenteditable="true" spellcheck="false"><p><br data-mce-bogus="1"></p></body></html>
```

**Comment by ashwinvis** [comment]
Ahhh...classic case of RTFD. Apologies and thank you. 😇

-------- Original Message --------

**Comment by Toshiro-Hitsugaya** [comment]
May be this will help you guys.
https://www.w3.org/TR/wai-aria-practices-1.1/examples/combobox/aria1.1pattern/listbox-combo.html

**Comment by The-Compiler** [comment]
@Toshiro-Hitsugaya how so?

**Comment by The-Compiler** [comment]
All of those (at least the ones I could test) are fixed via #4722.
